libs/core/include/core/bloom_filter.hpp

```diff
@@ -0,0 +1,224 @@
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/bitvector.hpp"
+
+#include <cstddef>
+#include <functional>
+#include <iterator>
+#include <utility>
+#include <vector>
+
+namespace fetch {
+
+namespace byte_array {
+class ConstByteArray;
+}
+
+namespace internal {
+
+class HashSource;
+
+/*
+ * An ordered collection of hash functions for generating pseudorandom
+ * std::size_t indices for the Bloom filter. To apply the functions,
+ * to an input, invoke the factory's operator() and use the resulting
+ * HashSource.
+ *
+ * The factory must be kept alive while its HashSource instances remain
+ * in use.
+ */
+class HashSourceFactory
+{
+public:
+  using Bytes     = fetch::byte_array::ConstByteArray;
+  using Function  = std::function<std::vector<std::size_t>(Bytes const &)>;
+  using Functions = std::vector<Function>;
+
+  explicit HashSourceFactory(Functions);
+  ~HashSourceFactory();
+
+  HashSourceFactory()                          = delete;
+  HashSourceFactory(HashSourceFactory const &) = delete;
+  HashSourceFactory(HashSourceFactory &&)      = delete;
+  HashSourceFactory &operator=(HashSourceFactory const &) = delete;
+  HashSourceFactory &operator=(HashSourceFactory &&) = delete;
+
+  HashSource operator()(Bytes const &) const;
+
+private:
+  Functions const hash_functions_;
+};
+
+/*
+ * Represents a sequential application of a HashSourceFactory's hash
+ * functions to a byte array. Outwardly it may be treated as an immutable,
+ * iterable collection of std::size_t.
+ *
+ * Not thread-safe. Not safe to use after the parent HashSourceFactory
+ * had been destroyed.
+ */
+class HashSource
+{
+public:
+  HashSource(HashSource &&) noexcept;
+  HashSource &operator=(HashSource &&) noexcept;
+  ~HashSource();
+
+  HashSource()                   = delete;
+  HashSource(HashSource const &) = delete;
+  HashSource &operator=(HashSource const &) = delete;
+
+  class HashSourceIterator
+  {
+  public:
+    using iterator_category = std::input_iterator_tag;
+    using value_type        = std::size_t const;
+    using difference_type   = std::ptrdiff_t;
+    using pointer           = std::size_t const *;
+    using reference         = std::size_t const &;
+
+    ~HashSourceIterator();
+
+    HashSourceIterator(HashSourceIterator const &);
+    HashSourceIterator(HashSourceIterator &&) noexcept;
+
+    HashSourceIterator &operator=(HashSourceIterator const &);
+    HashSourceIterator &operator=(HashSourceIterator &&) noexcept;
+
+    HashSourceIterator() = delete;
+
+    bool operator!=(HashSourceIterator const &other) const;
+    bool operator==(HashSourceIterator const &other) const;
+
+    HashSourceIterator const operator++(int);
```

Comment:
Do we really need a post-increment operator? I tend not to define them if I can help it.

Reply:
Fair.